From 11d8b2ed2141822d400c46232090c3dfbbf67198 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Fri, 13 Jun 2014 14:36:00 -0400
Subject: [block] Enable sysfs nomerge control for I/O requests in the plug list

Message-id: <1402670238-13916-73-git-send-email-snitzer@redhat.com>
Patchwork-id: 83884
O-Subject: [RHEL7.1 PATCH 072/150] block: Enable sysfs nomerge control for I/O requests in the plug list
Bugzilla: 1105204
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Jeff Moyer <jmoyer@redhat.com>

BZ: 1105204

Upstream commit 23779fbc99302dddab7f056ae47c3463169cbb64
Author: Alireza Haghdoost <alireza@cs.umn.edu>
Date:   Wed Oct 23 17:08:16 2013 +0100

    block: Enable sysfs nomerge control for I/O requests in the plug list

    This patch enables the sysfs to control I/O request merge
    functionality in the plug list. While this control has been
    implemented for the request queue, it was dismissed in the plug list.
    Therefore, block layer merges requests together (or attempt to merge)
    even if the merge capability was disable using sysfs nomerge parameter
    value 2.

    This limitation is directly affects functionality of io_submit()
    system call. The system call enables user to submit a bunch of IO
    requests from user space using struct iocb **ios input argument.
    However, the unconditioned merging functionality in the plug list
    potentially merges these requests together down the road. Therefore,
    there is no way to distinguish between an application sending bunch of
    sequential IOs and an application sending one big IO. Ultimately, all
    requests generated by the former app merge within the plug list
    together and looks similar to the second app.

    While the merging functionality is a desirable feature to improve the
    performance of IO subsystem for some applications, it is not useful
    for other application like ours at all.

    Signed-off-by: Alireza Haghdoost <alireza@cs.umn.edu>
    Reviewed-by: Jeff Moyer <jmoyer@redhat.com>

    Coding style modified.

    Signed-off-by: Jens Axboe <axboe@kernel.dk>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-core.c b/block/blk-core.c
index 7beec85..a4da95f 100644
--- a/block/blk-core.c
+++ b/block/blk-core.c
@@ -1486,6 +1486,9 @@ bool blk_attempt_plug_merge(struct request_queue *q, struct bio *bio,
  bool ret = false;
  struct list_head *plug_list;
 
+ if (blk_queue_nomerges(q))
+  goto out;
+
  plug = current->plug;
  if (!plug)
   goto out;
-- 
1.7.1